syntax = "proto3";

// REQUESTS
/*
Service receives a request containing a region, which
the service will return a list of databases currently
running in that region. (SERVER STREAMING)
*/
message DbListRequest {
  string region = 1;
}

/*
Service receives a request to login to specified database
with provided credentials. (UNARY)
*/
message DbAuthRequest {
  string databaseName = 1;
  string userName = 2;
  string password = 3;
}

/*
Service receives a payload of data to upload in chunks to
the specified database in the request (CLIENT STREAMING)
*/
message DbUploadRequest {
  string databaseName = 1;
  string databaseToken = 2;
  string payloadChunk = 3;
}


// RESPONSES
message DbListResponse {
  string databaseName = 1;
  string responseMessage = 4;
  int32 responseCode = 5;
}

message DbAuthResponse {
  string databaseToken = 1;
  string responseMessage = 2;
  int32 responseCode = 3;
}

message DbUploadResponse {
  string responseMessage = 1;
  int32 responseCode = 2;
}

// SERVICE
service DatabaseService {
  // SERVER STREAMING
  rpc listDatabases(DbListRequest) returns (stream DbListResponse);
  // UNARY
  rpc databaseLogin(DbAuthRequest) returns (DbAuthResponse);
  // CLIENT STREAMING
  rpc databaseUpload(stream DbUploadRequest) returns (DbUploadResponse);
}